File: sdks/php/src/Openrate.php
<?php

declare(strict_types=1);

namespace Openrate;

/**
 * openrate as a managed sidecar: the SDK spawns the `openrate` binary on a
 * loopback port, waits for it to answer /healthz, and terminates it at shutdown.
 * You never run a server by hand.
 *
 *   use Openrate\Openrate;
 *
 *   $base = Openrate::baseUrl();     // http://127.0.0.1:<port>
 *   $r = Openrate::convert('USD', 'ZAR', 100);
 *   echo $r['result'];
 *
 * This is the recommended mode for PHP. See README.md — php-fpm forks its
 * workers and the Go runtime inside libopenrate does not survive fork().
 */
final class Openrate
{
    public const VERSION = '0.1.2';

    /** @var resource|null */
    private static $proc = null;

    /** @var string|null */
    private static $base = null;

    /** @var bool */
    private static $shutdownRegistered = false;

    /**
     * Start the sidecar (idempotent). Returns the base URL (http://host:port).
     *
     * @param array{port?:int,base?:string,sources?:string,refresh?:string,ui?:bool,ratelimit?:int,env?:array<string,string>,timeout?:float} $opts
     */
    public static function start(array $opts = []): string
    {
        if (self::running()) {
            return self::$base;
        }

        $port = $opts['port'] ?? self::freePort();
        $addr = "127.0.0.1:{$port}";

        $env = self::inheritedEnv();
        $env['OPENRATE_ADDR'] = $addr;
        // The binary defaults to 120 API requests/minute per IP, which is
        // anti-scraping for a PUBLIC deployment. This sidecar is on loopback and
        // serves exactly one client — us — and that budget is small enough that
        // the SDK's own startup health polling can exhaust it and hand the first
        // real call an HTTP 429. Default it off here and let a caller who wants
        // it back pass ['ratelimit' => 120].
        $env['OPENRATE_RATELIMIT'] = (string) ($opts['ratelimit'] ?? 0);
        if (isset($opts['base'])) {
            $env['OPENRATE_BASE'] = $opts['base'];
        }
        if (isset($opts['sources'])) {
            $env['OPENRATE_SOURCES'] = $opts['sources'];
        }
        if (isset($opts['refresh'])) {
            $env['OPENRATE_REFRESH'] = $opts['refresh'];
        }
        if (isset($opts['ui'])) {
            $env['OPENRATE_UI'] = $opts['ui'] ? 'true' : 'false';
        }
        if (isset($opts['env'])) {
            $env = array_merge($env, $opts['env']);
        }

        $descriptors = [0 => STDIN, 1 => STDOUT, 2 => STDERR];

        $proc = proc_open([self::binaryPath()], $descriptors, $pipes, null, $env);
        if (!\is_resource($proc)) {
            throw new OpenrateException('failed to spawn the openrate binary');
        }
        self::$proc = $proc;
        self::$base = "http://{$addr}";

        try {
            self::waitHealthy(self::$base, (float) ($opts['timeout'] ?? 15.0));
        } catch (\Throwable $e) {
            self::stop();
            throw $e;
        }

        if (!self::$shutdownRegistered) {
            register_shutdown_function([self::class, 'stop']);
            self::$shutdownRegistered = true;
        }

        return self::$base;
    }

    /** The running base URL, starting the sidecar if needed. */
    public static function baseUrl(): string
    {
        return self::running() ? self::$base : self::start();
    }

    /** The JSON API base (…/api/v1). */
    public static function apiBaseUrl(): string
    {
        return self::baseUrl() . '/api/v1';
    }

    /**
     * GET /api/v1/convert.
     *
     * @return array<string,mixed>
     */
    public static function convert(string $from, string $to, float $amount = 1.0): array
    {
        return self::get('/convert?' . http_build_query([
            'from' => $from,
            'to' => $to,
            'amount' => $amount,
        ]));
    }

    /**
     * GET /api/v1/rates.
     *
     * @return array<string,mixed>
     */
    public static function rates(?string $base = null): array
    {
        return self::get('/rates' . ($base !== null ? '?' . http_build_query(['base' => $base]) : ''));
    }

    /**
     * GET /api/v1/meta.
     *
     * @return array<string,mixed>
     */
    public static function meta(): array
    {
        return self::get('/meta');
    }

    /**
     * One GET against the JSON API. $path is relative to /api/v1.
     *
     * @return array<string,mixed>
     */
    public static function get(string $path): array
    {
        $url = self::apiBaseUrl() . $path;
        $ctx = stream_context_create(['http' => ['timeout' => 30, 'ignore_errors' => true]]);
        $body = @file_get_contents($url, false, $ctx);
        if ($body === false) {
            throw new OpenrateException("GET {$path} failed");
        }
        $status = self::parseStatus(self::lastResponseHeaders($http_response_header ?? null) ?? []);
        $decoded = json_decode($body, true);
        if ($status !== 200) {
            $message = \is_array($decoded) && isset($decoded['error'])
                ? (string) $decoded['error']
                : substr($body, 0, 200);
            throw new OpenrateException("GET {$path}: HTTP {$status}: {$message}");
        }
        if (!\is_array($decoded)) {
            throw new OpenrateException("GET {$path}: response was not a JSON object");
        }

        return $decoded;
    }

    /** Stop the sidecar if running. Idempotent. */
    public static function stop(): void
    {
        if (\is_resource(self::$proc)) {
            $status = proc_get_status(self::$proc);
            if ($status['running']) {
                proc_terminate(self::$proc, \defined('SIGTERM') ? SIGTERM : 15);
                $deadline = microtime(true) + 5.0;
                while (microtime(true) < $deadline) {
                    if (!proc_get_status(self::$proc)['running']) {
                        break;
                    }
                    usleep(50_000);
                }
                if (proc_get_status(self::$proc)['running']) {
                    proc_terminate(self::$proc, \defined('SIGKILL') ? SIGKILL : 9);
                }
            }
            proc_close(self::$proc);
        }
        self::$proc = null;
        self::$base = null;
    }

    // ---------------------------------------------------------------- internals

    private static function running(): bool
    {
        if (!\is_resource(self::$proc)) {
            return false;
        }

        return (bool) proc_get_status(self::$proc)['running'];
    }

    private static function binaryPath(): string
    {
        $env = getenv('OPENRATE_BINARY');
        if ($env !== false && $env !== '') {
            return $env;
        }

        $name = self::isWindows() ? 'openrate.exe' : 'openrate';
        $bundled = \dirname(__DIR__) . DIRECTORY_SEPARATOR . 'bin' . DIRECTORY_SEPARATOR . $name;
        if (is_file($bundled)) {
            return $bundled;
        }

        $found = self::which('openrate');
        if ($found !== null) {
            return $found;
        }

        throw new OpenrateException(
            'openrate binary not found. Set OPENRATE_BINARY, install a platform ' .
            'package, or build it: `go build -o sdks/php/bin/openrate ./cmd/openrate`'
        );
    }

    private static function which(string $cmd): ?string
    {
        $exts = self::isWindows() ? explode(';', (string) getenv('PATHEXT')) : [''];
        foreach (explode(PATH_SEPARATOR, (string) getenv('PATH')) as $dir) {
            foreach ($exts as $ext) {
                $candidate = $dir . DIRECTORY_SEPARATOR . $cmd . $ext;
                if (is_file($candidate) && is_executable($candidate)) {
                    return $candidate;
                }
            }
        }

        return null;
    }

    private static function isWindows(): bool
    {
        return \PHP_OS_FAMILY === 'Windows';
    }

    /** @return array<string,string> */
    private static function inheritedEnv(): array
    {
        $env = [];
        foreach ($_ENV as $k => $v) {
            $env[(string) $k] = (string) $v;
        }
        if ($env === []) {
            foreach (\array_keys(getenv()) as $k) {
                $env[(string) $k] = (string) getenv((string) $k);
            }
        }

        return $env;
    }

    private static function freePort(): int
    {
        $sock = @stream_socket_server('tcp://127.0.0.1:0', $errno, $errstr);
        if ($sock === false) {
            throw new OpenrateException("could not allocate a free port: {$errstr}");
        }
        $name = (string) stream_socket_get_name($sock, false);
        fclose($sock);

        return (int) substr($name, (int) strrpos($name, ':') + 1);
    }

    private static function waitHealthy(string $base, float $timeout): void
    {
        $deadline = microtime(true) + $timeout;
        $last = 'connection refused';
        $ctx = stream_context_create(['http' => ['timeout' => 1, 'ignore_errors' => true]]);
        while (microtime(true) < $deadline) {
            $body = @file_get_contents($base . '/healthz', false, $ctx);
            $headers = self::lastResponseHeaders($http_response_header ?? null);
            if ($body !== false && $headers !== null) {
                $status = self::parseStatus($headers);
                if ($status === 200) {
                    return;
                }
                $last = "status {$status}";
            }
            // 200 ms, not 50 ms: /healthz goes through the same per-IP limiter
            // as the API, so a tight poll spends a budget the caller wanted.
            usleep(200_000);
        }
        throw new OpenrateException("openrate did not become healthy within {$timeout}s: {$last}");
    }

    /**
     * PHP 8.4 deprecated the magic locally-scoped $http_response_header in
     * favour of http_get_last_response_headers().
     *
     * @param array<int,string>|null $magic
     * @return array<int,string>|null
     */
    private static function lastResponseHeaders(?array $magic): ?array
    {
        if (\function_exists('http_get_last_response_headers')) {
            return \http_get_last_response_headers();
        }

        return $magic;
    }

    /** @param array<int,string> $headers */
    private static function parseStatus(array $headers): int
    {
        foreach ($headers as $h) {
            if (preg_match('#^HTTP/\S+\s+(\d+)#', $h, $m)) {
                return (int) $m[1];
            }
        }

        return 0;
    }
}

File: sdks/dotnet/OpenRate.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Sockets;
using System.Runtime.InteropServices;
using System.Threading;

namespace OpenRate
{
    /// <summary>
    /// openrate as a <b>child process</b> on 127.0.0.1, spoken to over HTTP.
    ///
    /// <para><b>The recommended default for .NET.</b> No native library, no
    /// unsafe code, no platform matrix — it works on Windows, where the shared
    /// library does not exist.</para>
    ///
    /// <code>
    /// using var rates = Sidecar.Start(new Sidecar.Options { Base = "ZAR" });
    /// string converted = rates.Convert("USD", "ZAR", 100);
    /// </code>
    ///
    /// <para><b>This fetches.</b> <c>openrate serve</c> starts its refresher at
    /// startup, so standing one up means outbound requests. If "no packets
    /// unless I ask" is the property you want, that is
    /// <see cref="OpenRateEngine"/> on the direct path, and it is a guarantee
    /// rather than a setting.</para>
    ///
    /// <para>Unlike llmux's .NET SDK this is <b>not a process-wide
    /// singleton</b>: you get an instance you own and dispose. Several servers
    /// with different bases or source sets is a normal thing to want from a
    /// rates service, and a static singleton would forbid it.</para>
    /// </summary>
    public sealed class Sidecar : IDisposable
    {
        public const string Version = "0.1.2";

        public sealed class Options
        {
            /// <summary>Fixed port; defaults to an ephemeral free port.</summary>
            public int? Port { get; set; }
            /// <summary>Default presentation base currency (server default: ZAR).</summary>
            public string? Base { get; set; }
            /// <summary>Comma-separated FX sources; empty means openrate's defaults.</summary>
            public string? Sources { get; set; }
            /// <summary>Serve the embedded web console. Off: an API client does not need it.</summary>
            public bool Ui { get; set; }
            /// <summary>Extra environment for the child process.</summary>
            public IDictionary<string, string>? Env { get; set; }
            /// <summary>How long to wait for /healthz (default 30s).</summary>
            public TimeSpan? Timeout { get; set; }
            /// <summary>
            /// Also wait until the server actually holds rates, not merely until
            /// it is listening. Default true, and it is the difference between a
            /// working client and a mystery.
            ///
            /// <para><c>/healthz</c> is a LIVENESS probe: openrate answers 200
            /// the moment it binds, while its first fetch is still in flight,
            /// and every conversion in that window returns
            /// <c>{"error":"unknown or unreachable currency pair"}</c> — which
            /// reads as a wrong currency code, not as "not ready yet".</para>
            /// </summary>
            public bool WaitForRates { get; set; } = true;
            /// <summary>How long to wait for the first rates (default 60s).</summary>
            public TimeSpan? RatesTimeout { get; set; }
        }

        private readonly Process _proc;
        private readonly HttpClient _http;
        private readonly EventHandler _exitHandler;
        private int _disposed;

        /// <summary>http://127.0.0.1:&lt;port&gt;</summary>
        public string BaseUrl { get; }

        private Sidecar(Process proc, string baseUrl)
        {
            _proc = proc;
            BaseUrl = baseUrl;
            _http = new HttpClient { Timeout = TimeSpan.FromSeconds(30) };
            _exitHandler = (_, _) => Terminate();
            AppDomain.CurrentDomain.ProcessExit += _exitHandler;
        }

        /// <summary>Spawn openrate and wait for it to be ready.</summary>
        public static Sidecar Start(Options? opts = null)
        {
            opts ??= new Options();
            int port = opts.Port ?? FreePort();
            string addr = $"127.0.0.1:{port}";

            var psi = new ProcessStartInfo
            {
                FileName = BinaryPath(),
                UseShellExecute = false,
            };
            psi.ArgumentList.Add("-addr");
            psi.ArgumentList.Add(addr);
            psi.ArgumentList.Add("-ui=" + (opts.Ui ? "true" : "false"));
            if (opts.Base != null) { psi.Environment["OPENRATE_BASE"] = opts.Base; }
            if (opts.Sources != null) { psi.Environment["OPENRATE_SOURCES"] = opts.Sources; }
            if (opts.Env != null)
            {
                foreach (var kv in opts.Env) { psi.Environment[kv.Key] = kv.Value; }
            }

            Process proc;
            try
            {
                proc = Process.Start(psi)
                    ?? throw new OpenRateException("failed to spawn the openrate binary");
            }
            catch (Exception e) when (e is not OpenRateException)
            {
                throw new OpenRateException("failed to spawn the openrate binary", e);
            }

            var sidecar = new Sidecar(proc, $"http://{addr}");
            try
            {
                sidecar.WaitHealthy(opts.Timeout ?? TimeSpan.FromSeconds(30));
                if (opts.WaitForRates)
                {
                    sidecar.WaitForRates(opts.RatesTimeout ?? TimeSpan.FromSeconds(60));
                }
            }
            catch
            {
                // Never leave a child process behind because startup failed.
                sidecar.Dispose();
                throw;
            }
            return sidecar;
        }

        /// <summary>GET /api/v1/meta.</summary>
        public string Meta() => Get("/api/v1/meta");

        /// <summary>GET /api/v1/rates?base=…</summary>
        public string Rates(string baseCurrency) =>
            Get("/api/v1/rates?base=" + Uri.EscapeDataString(baseCurrency));

        /// <summary>GET /api/v1/convert?from=…&amp;to=…&amp;amount=…</summary>
        public string Convert(string from, string to, double amount = 1) =>
            Get($"/api/v1/convert?from={Uri.EscapeDataString(from)}"
                + $"&to={Uri.EscapeDataString(to)}"
                + $"&amount={amount.ToString(System.Globalization.CultureInfo.InvariantCulture)}");

        /// <summary>One GET against the running server; the body, whatever the status.</summary>
        public string Get(string path)
        {
            try
            {
                return _http.GetStringAsync(BaseUrl + path).GetAwaiter().GetResult();
            }
            catch (HttpRequestException e)
            {
                // openrate answers 4xx with a JSON error document, which is a
                // result rather than an exception. Hand it back.
                if (e.StatusCode != null)
                {
                    using var resp = _http.GetAsync(BaseUrl + path).GetAwaiter().GetResult();
                    return resp.Content.ReadAsStringAsync().GetAwaiter().GetResult();
                }
                throw new OpenRateException($"GET {path} failed", e);
            }
        }

        /// <summary>Stop the child process. Idempotent — use <c>using</c>.</summary>
        public void Dispose()
        {
            if (Interlocked.Exchange(ref _disposed, 1) == 1)
            {
                return;
            }
            Terminate();
            AppDomain.CurrentDomain.ProcessExit -= _exitHandler;
            _http.Dispose();
        }

        private void Terminate()
        {
            try
            {
                if (!_proc.HasExited)
                {
                    _proc.Kill(entireProcessTree: true);
                    _proc.WaitForExit(5000);
                }
            }
            catch
            {
                // Best effort: the process may already be gone.
            }
        }

        private void WaitHealthy(TimeSpan timeout)
        {
            DateTime deadline = DateTime.UtcNow + timeout;
            string last = "connection refused";
            while (DateTime.UtcNow < deadline)
            {
                if (_proc.HasExited)
                {
                    throw new OpenRateException(
                        $"the openrate binary exited before becoming healthy (status {_proc.ExitCode})");
                }
                try
                {
                    using HttpResponseMessage resp =
                        _http.GetAsync(BaseUrl + "/healthz").GetAwaiter().GetResult();
                    if (resp.StatusCode == HttpStatusCode.OK)
                    {
                        return;
                    }
                    last = $"status {(int)resp.StatusCode}";
                }
                catch (Exception e)
                {
                    last = e.Message;
                }
                Thread.Sleep(100);
            }
            throw new OpenRateException($"openrate did not become healthy within {timeout}: {last}");
        }

        /// <summary>
        /// Poll /api/v1/meta until the currency list is non-empty. See
        /// <see cref="Options.WaitForRates"/>: this is a readiness check built
        /// on a liveness endpoint, because openrate publishes no readiness
        /// endpoint of its own. The direct path's refresher does — its
        /// <c>ready</c> method.
        /// </summary>
        private void WaitForRates(TimeSpan timeout)
        {
            DateTime deadline = DateTime.UtcNow + timeout;
            string last = "no rates yet";
            while (DateTime.UtcNow < deadline)
            {
                if (_proc.HasExited)
                {
                    throw new OpenRateException("the openrate binary exited while fetching its first rates");
                }
                try
                {
                    if (HasCurrencies(Get("/api/v1/meta")))
                    {
                        return;
                    }
                    last = "the currency list is still empty";
                }
                catch (Exception e)
                {
                    last = e.Message;
                }
                Thread.Sleep(250);
            }
            throw new OpenRateException(
                $"openrate is listening but holds no rates after {timeout} ({last}). "
                + "Every source fetch failed — the usual cause is no outbound network. "
                + "Set Options.WaitForRates = false to start anyway.");
        }

        /// <summary>
        /// True if the meta document's <c>currencies</c> array has anything in
        /// it. Deliberately not a JSON parse: this SDK has no JSON dependency,
        /// and "is this array empty" does not need one.
        /// </summary>
        internal static bool HasCurrencies(string metaJson)
        {
            int at = metaJson.IndexOf("\"currencies\"", StringComparison.Ordinal);
            if (at < 0) { return false; }
            int open = metaJson.IndexOf('[', at);
            if (open < 0) { return false; }
            for (int i = open + 1; i < metaJson.Length; i++)
            {
                char c = metaJson[i];
                if (c == ']') { return false; }
                if (!char.IsWhiteSpace(c)) { return true; }
            }
            return false;
        }

        // ---------------------------------------------------------------- lookup

        private static string BinaryPath()
        {
            string? env = Environment.GetEnvironmentVariable("OPENRATE_BINARY");
            if (!string.IsNullOrEmpty(env)) { return env!; }

            bool windows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
            string name = windows ? "openrate.exe" : "openrate";
            string bundled = Path.Combine(AppContext.BaseDirectory, "bin", name);
            if (File.Exists(bundled)) { return bundled; }

            string? found = Which(name);
            if (found != null) { return found; }

            throw new OpenRateException(
                "openrate binary not found. Set OPENRATE_BINARY, or build it: "
                + "`go build -o sdks/dotnet/bin/openrate ./cmd/openrate`");
        }

        private static string? Which(string cmd)
        {
            string? path = Environment.GetEnvironmentVariable("PATH");
            if (path == null) { return null; }
            foreach (string dir in path.Split(Path.PathSeparator))
            {
                string candidate = Path.Combine(dir, cmd);
                if (File.Exists(candidate)) { return candidate; }
            }
            return null;
        }

        private static int FreePort()
        {
            var listener = new TcpListener(IPAddress.Loopback, 0);
            listener.Start();
            int port = ((IPEndPoint)listener.LocalEndpoint).Port;
            listener.Stop();
            return port;
        }
    }

    /// <summary>Thrown when openrate cannot be located, started, or made to answer.</summary>
    public sealed class OpenRateException : Exception
    {
        public OpenRateException(string message) : base(message) { }
        public OpenRateException(string message, Exception inner) : base(message, inner) { }
    }
}
